File: node/tests/wire/vbp/test_auth.js
'use strict';

const { test } = require('node:test');
const assert = require('node:assert');
const { SCRAMClient, plainClientFirst, VBPAuthError } = require('../../../src/wire/vbp/auth');

// ---------------------------------------------------------------------------
// PLAIN
// ---------------------------------------------------------------------------

test('auth PLAIN: format is \\0user\\0pass', () => {
  const buf = plainClientFirst('alice', 'secret');
  // \x00 alice \x00 secret
  assert.strictEqual(buf[0], 0);
  assert.strictEqual(buf.subarray(1, 6).toString('utf-8'), 'alice');
  assert.strictEqual(buf[6], 0);
  assert.strictEqual(buf.subarray(7).toString('utf-8'), 'secret');
});

test('auth PLAIN: works with non-ASCII passwords (utf-8)', () => {
  const buf = plainClientFirst('u', 'пароль');
  // \x00 u \x00 + utf8
  assert.strictEqual(buf[0], 0);
  assert.strictEqual(buf[2], 0);
});

// ---------------------------------------------------------------------------
// SCRAM
// ---------------------------------------------------------------------------

test('auth SCRAM: clientFirstBare has n=user,r=nonce', () => {
  const c = new SCRAMClient('alice', 'pw');
  const cfb = c.clientFirstBare();
  assert.ok(cfb.startsWith('n=alice,r='));
  // nonce is 18 bytes base64 = 24 chars
  const m = cfb.match(/r=([A-Za-z0-9+/=]+)$/);
  assert.ok(m);
  assert.strictEqual(m[1].length, 24);
});

test('auth SCRAM: saslname escapes = and ,', () => {
  const c = new SCRAMClient('a=b,c', 'pw');
  const cfb = c.clientFirstBare();
  assert.ok(cfb.startsWith('n=a=3Db=2Cc,'));
});

test('auth SCRAM: clientFinalMessage verifies server-first nonce', () => {
  const c = new SCRAMClient('alice', 'pw');
  const cfb = c.clientFirstBare();
  const clientNonce = cfb.match(/r=([^,]+)$/)[1];
  const serverFirst = `r=${clientNonce}AAAAAAAAAAAAAAAAAAAAAA,s=c2FsdA==,i=4096`;
  const finalMsg = c.clientFinalMessage(serverFirst);
  // Format: c=<b64>,r=<nonce>,p=<b64>
  assert.ok(finalMsg.startsWith('c='));
  const parts = finalMsg.split(',');
  assert.strictEqual(parts.length, 3);
  assert.ok(parts[1].startsWith('r='));
  assert.ok(parts[2].startsWith('p='));
  // The nonce in r= should equal combined nonce (= client nonce + AAA...)
  assert.strictEqual(parts[1].slice(2), clientNonce + 'AAAAAAAAAAAAAAAAAAAAAA');
});

test('auth SCRAM: rejects server-first that does not begin with client nonce', () => {
  const c = new SCRAMClient('alice', 'pw');
  const serverFirst = 'r=DIFFERENTNONCE,s=c2FsdA==,i=4096';
  assert.throws(() => c.clientFinalMessage(serverFirst), VBPAuthError);
});

test('auth SCRAM: rejects malformed server-first', () => {
  const c = new SCRAMClient('alice', 'pw');
  assert.throws(() => c.clientFinalMessage('garbage'), VBPAuthError);
  assert.throws(() => c.clientFinalMessage('r=x,s=y'), VBPAuthError); // missing i
});

test('auth SCRAM: cbind_input is exactly the gs2_header (RFC 5802 §6 pencil vector)', () => {
  // Per RFC 5802 §6, when the gs2-flag is 'n' (no channel binding),
  // cbind-data is ABSENT, so cbind-input is just the gs2-header.
  // The canonical pencil test vector is c=biws, which is
  // base64("n,,"). Anything else (e.g. "n,,n=user,r=nonce") is a
  // conformance bug — that would imply a non-'n' gs2-flag.
  const c = new SCRAMClient('alice', 'pw');
  const serverFirst = `r=${c.clientNonce}AAAA,s=c2FsdA==,i=4096`;
  const finalMsg = c.clientFinalMessage(serverFirst);
  // finalMsg format: c=<b64>,r=<nonce>,p=<b64>
  const parts = finalMsg.split(',');
  assert.strictEqual(parts[0].slice(0, 2), 'c=');
  const cPart = parts[0].slice(2);
  // The exact pencil vector from RFC 5802 §5 / §6: c=biws.
  assert.strictEqual(cPart, 'biws',
    `cbind channel-binding should be base64("n,,") = "biws", got ${JSON.stringify(cPart)}`);
  // Defensive: decode and assert the literal string.
  assert.strictEqual(Buffer.from(cPart, 'base64').toString('utf-8'), 'n,,');
});

test('auth SCRAM: empty username throws', () => {
  assert.throws(() => new SCRAMClient('', 'pw'), VBPAuthError);
});

test('auth SCRAM: verifyServerFinal with no v= is a no-op (dev mode)', () => {
  const c = new SCRAMClient('alice', 'pw');
  const serverFirst = `r=${c.clientNonce}AAAA,s=c2FsdA==,i=4096`;
  c.clientFinalMessage(serverFirst);
  // Empty body (no v=) — dev server skips signature verification.
  c.verifyServerFinal('');
  // Garbage v= should throw a real signature mismatch.
  assert.throws(() => c.verifyServerFinal('v=AQIDBAUGBwgJCgsMDQ4PEA'), VBPAuthError);
});

File: python/tests/wire/vbp/test_auth.py
"""Tests for the vbp.wire.auth module — PLAIN body and SCRAM-SHA-256 client."""
from __future__ import annotations

import base64
import hashlib
import hmac
import os
import sys
import unittest

# Make sure we can import the vedadb package from the repo root.
HERE = os.path.dirname(os.path.abspath(__file__))
REPO_ROOT = os.path.abspath(os.path.join(HERE, "..", "..", ".."))
if REPO_ROOT not in sys.path:
    sys.path.insert(0, REPO_ROOT)

from vedadb.wire.vbp.auth import (  # noqa: E402
    SCRAMClient,
    VBPAuthError,
    plain_client_first,
)


class TestPLAINBody(unittest.TestCase):
    def test_plain_format(self):
        body = plain_client_first("admin", "secret")
        # SASL PLAIN: [authzid] NUL authcid NUL password.  v1 uses no authzid.
        self.assertEqual(body, b"\x00admin\x00secret")

    def test_plain_unicode(self):
        body = plain_client_first("user", "héllo")
        self.assertEqual(body, b"\x00user\x00h\xc3\xa9llo")


class TestSCRAMClientNonce(unittest.TestCase):
    def test_client_nonce_is_ascii_printable(self):
        scram = SCRAMClient("admin", "TestPassword123!")
        cn = scram.client_nonce
        # Base64 alphabet is ASCII-printable.
        for ch in cn:
            self.assertTrue(ch.isascii(), f"non-ASCII char in client nonce: {ch!r}")
            self.assertLessEqual(ord(ch), 0x7F, f"non-printable char in client nonce: {ch!r}")

    def test_client_first_message(self):
        scram = SCRAMClient("admin", "TestPassword123!")
        msg = scram.client_first()
        self.assertTrue(msg.startswith("n=admin,r="))
        self.assertEqual(msg.split(",r=")[1], scram.client_nonce)


class TestSCRAMClientFirstBare(unittest.TestCase):
    def test_username_saslname_escaping(self):
        scram = SCRAMClient("user=name", "pw")
        msg = scram.client_first()
        # "=" should be escaped as "=3D".
        self.assertIn("n=user=3Dname", msg)


class TestSCRAMClientFinal(unittest.TestCase):
    def test_client_final_message(self):
        scram = SCRAMClient("admin", "TestPassword123!")
        cf = scram.client_first()
        # Hand-rolled server-first for deterministic testing.
        salt = os.urandom(16)
        salt_b64 = base64.b64encode(salt).decode("ascii")
        iters = 4096
        # Combine nonces: client_nonce + server_nonce.
        server_nonce = "ABCDEFGH"  # ascii, valid
        combined_nonce = scram.client_nonce + server_nonce
        server_first = f"r={combined_nonce},s={salt_b64},i={iters}"

        final = scram.client_final_message(server_first)
        # Format: c=<base64-cbind>,r=<combined>,p=<base64-proof>
        parts = final.split(",")
        self.assertEqual(len(parts), 3)
        self.assertTrue(parts[0].startswith("c="))
        self.assertTrue(parts[1].startswith("r="))
        self.assertEqual(parts[1], f"r={combined_nonce}")
        self.assertTrue(parts[2].startswith("p="))

    def test_client_final_wrong_nonce_prefix_rejected(self):
        scram = SCRAMClient("admin", "pw")
        # Server-first where the nonce doesn't start with the client nonce.
        # This signals a MITM.
        scram.client_first()
        with self.assertRaises(VBPAuthError) as cm:
            scram.client_final_message("r=NOT_CLIENT,s=" + base64.b64encode(b"\x00" * 16).decode() + ",i=4096")
        self.assertIn("MITM", cm.exception.message)

    def test_client_final_bad_iter_rejected(self):
        scram = SCRAMClient("admin", "pw")
        cn = scram.client_nonce
        with self.assertRaises(VBPAuthError):
            scram.client_final_message(f"r={cn}xx,s={base64.b64encode(b'x' * 16).decode()},i=NOTANUMBER")


class TestSCRAMServerSignatureVerify(unittest.TestCase):
    """Verify the server signature is computed correctly per RFC 5802."""

    def test_signature_round_trip(self):
        # Build a SCRAMClient and a fake server that computes the
        # expected server signature independently.
        scram = SCRAMClient("admin", "TestPassword123!", simplified=False)
        cf = scram.client_first()
        cn = scram.client_nonce
        salt = b"1234567890123456"  # 16 bytes
        salt_b64 = base64.b64encode(salt).decode("ascii")
        iters = 4096
        server_nonce = "SRVNONCE"
        combined = cn + server_nonce
        server_first = f"r={combined},s={salt_b64},i={iters}"

        # Force cache of salt+iters before computing final.
        scram._salt_b64 = salt_b64
        scram._iters = iters

        # Build the client-final message.
        final = scram.client_final_message(server_first)

        # Independently compute the expected server signature.
        salted_password = hashlib.pbkdf2_hmac(
            "sha256", b"TestPassword123!", salt, iters
        )
        server_key = hmac.new(salted_password, b"Server Key", hashlib.sha256).digest()
        # Extract p= from final.
        p_part = final.rsplit(",p=", 1)[1]
        client_proof = base64.b64decode(p_part)
        # Re-derive stored_key and client_key from client_proof.
        client_key = hmac.new(salted_password, b"Client Key", hashlib.sha256).digest()
        stored_key = hashlib.sha256(client_key).digest()
        client_signature = hmac.new(stored_key, scram.auth_message, hashlib.sha256).digest()
        # client_proof = client_key XOR client_signature
        self.assertEqual(
            client_proof,
            bytes(a ^ b for a, b in zip(client_key, client_signature)),
        )
        # server_signature = HMAC(server_key, auth_message)
        expected_server_sig = hmac.new(server_key, scram.auth_message, hashlib.sha256).digest()
        server_final = f"v={base64.b64encode(expected_server_sig).decode('ascii')}"
        # This should verify.
        scram.verify_server_final(server_final)

    def test_signature_mismatch_rejected(self):
        scram = SCRAMClient("admin", "TestPassword123!", simplified=False)
        cn = scram.client_nonce
        salt = b"1234567890123456"
        salt_b64 = base64.b64encode(salt).decode("ascii")
        iters = 4096
        combined = cn + "SRV"
        scram.client_first()
        scram.client_final_message(f"r={combined},s={salt_b64},i={iters}")
        scram._salt_b64 = salt_b64
        scram._iters = iters
        # A bogus server signature.
        bogus = base64.b64encode(b"x" * 32).decode("ascii")
        with self.assertRaises(VBPAuthError) as cm:
            scram.verify_server_final(f"v={bogus}")
        self.assertIn("MITM", cm.exception.message)

    def test_signature_wrong_prefix_rejected(self):
        scram = SCRAMClient("admin", "pw")
        cn = scram.client_nonce
        salt_b64 = base64.b64encode(b"x" * 16).decode("ascii")
        iters = 4096
        combined = cn + "SRV"
        scram.client_first()
        scram.client_final_message(f"r={combined},s={salt_b64},i={iters}")
        scram._salt_b64 = salt_b64
        scram._iters = iters
        # Server-final with 'e=' (error) prefix.
        with self.assertRaises(VBPAuthError):
            scram.verify_server_final("e=invalid-server-signature")


if __name__ == "__main__":
    unittest.main()
